File: Data Structures & Algorithms/house-robber-ii/submission-0.java
class Solution {
    // Time Complexity: O(N)
    // Space Complexity: O(1)
    public int rob(int[] nums) {
        if (nums.length == 1) return nums[0];
        if (nums.length == 2) return Math.max(nums[0], nums[1]);
        int prev2 = 0;
        int prev1 = nums[0];
        int dp1 = 0;
        

        for (int i = 1; i < nums.length - 1; i++) {
            dp1 = Math.max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = dp1;
        }

        prev1 = nums[1];
        prev2 = 0;
        int dp2 = 0;

        for (int i = 2; i < nums.length; i++) {
            dp2 = Math.max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = dp2;
        }

        return Math.max(dp1, dp2);
    }
}

File: Data Structures & Algorithms/sliding-window-maximum/submission-0.py
# Initial approach:
# Maintain a heap with k capacity
# Time complexity: O(nlogk)
# Space complexity: O(k)


class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        maxheap = []
        result = []

        for i in range(k):
            heapq.heappush(maxheap, (-nums[i], i))
        result.append(-maxheap[0][0])
        
        resultindex = k

        while resultindex < len(nums):
            heapq.heappush(maxheap, (-nums[resultindex], resultindex))

            while maxheap and maxheap[0][1] <= resultindex - k:
                heapq.heappop(maxheap)
            
            result.append(-maxheap[0][0])
            resultindex += 1
        
        return result
            


